keps/sig-api-machinery/3396-websocket-protocol-enhancements/README.md

```diff
@@ -0,0 +1,867 @@
+
+kubectl commands `exec`, `cp`, `attach`, and `port-forward` currently use [SPDY](https://en.wikipedia.org/wiki/SPDY)
+to communicate with API server.
+[SPDY is deprecated and its use is causing issues](https://github.com/kubernetes/kubernetes/issues/7452).
+API server already supports WebSocket for those API endpoints and the plan is to [migrate the commands to
+WebSocket protocol](https://github.com/kubernetes/kubernetes/issues/89163). Unfortunately, the application-level
+protocol that API server implements on top of WebSocket has
+[deficiencies](https://github.com/kubernetes/kubernetes/issues/89899) that block the transition.
+This KEP aims to resolve them to unblock the migration and to allow all API consumers to use WebSocket rather than SPDY.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+API server defines an application-level protocol on top of WebSocket connection. It's a framing protocol
+that allows to multiplex several logical bidirectional channels on top of a single WebSocket connection.
+
+Deficiencies in the current (v4) WebSocket protocol:
+
+- No mechanism to do a half-close of a logical channel. Neither client nor server can indicate that no
+  more data is coming on a particular channel. This makes it impossible to implement certain use-cases in those
+  kubectl commands. For example, `kubectl cp file pod` spawns `tar` in the target pod and streams the contents of a file
+  as a tar stream via STDIN to the `tar` command. It needs to close STDIN to indicate that all the data has been sent.
+  If it cannot do it, `tar` gets stuck forever, waiting for more data.
+
+- No mechanism to do a reset of a logical channel to stop the other party from sending more data. This is needed for
+  parity with SPDY and is used mostly in error handling situations.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Define a new WebSocket protocol version (v5) without the deficiencies described above.
+- Change kubectl to use new WebSocket protocol for `exec`, `cp`, and `attach`. It should fall back to SPDY when
+  new protocol is not supported by the server.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Removal of SPDY in any part of the system. In particular, API server will still have to support SPDY for some time
+  for backwards compatibility with older clients.
+- Moving `port-forward` to WebSocket. It's not clear how `port-forward` via WebSocket should work. This can be done
+  separately later.
```

Comment:
I'm not against making `port-forward` part of this proposal. I don't know if it is supported by the server at the moment. If we think head-of-line blocking is not an issue we care about then it shouldn't be too hard to implement on top of the same protocol.

Reply:
I think this is a good choice. We might want to do something funky that's not a WebSocket.

Reply:
Port forward should be supported by the server. https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cri/streaming/portforward/websocket.go

What is the concern with head of line blocking?

Reply:
Removed exclusion of port-forward.